src/main/scala/com/databricks/spark/avro/AvroRelation.scala

```diff
@@ -106,15 +135,28 @@ case class AvroRelation(location: String)(@transient val sqlContext: SQLContext)
           ArrayType(schemaType.dataType, containsNull = schemaType.nullable),
           nullable = false)
 
-      case UNION => avroSchema.getTypes.toSeq match {
-        case Seq(t1, t2) if t1.getType == NULL => toSqlType(t2).copy(nullable = true)
-        case Seq(t1, t2) if t2.getType == NULL => toSqlType(t1).copy(nullable = true)
-        case other =>
-          sys.error(s"Union types with anything other than null not supported: $other")
+      case MAP =>
+        val schemaType = toSqlType(avroSchema.getValueType)
+        SchemaType(
+          MapType(StringType, schemaType.dataType, valueContainsNull = schemaType.nullable),
+          nullable = false)
+
+      case UNION => {
+        if (avroSchema.getTypes.map(x => x.getType).contains(NULL)) {
+          // In case of a union with null, eliminate it and make a recursive call
+          toSqlType(Schema.createUnion(
+            avroSchema.getTypes.filterNot(elm => elm.getType == NULL))).copy(nullable = true)
+        } else avroSchema.getTypes.toSeq match {
+          case Seq(t1, t2) if Set(t2.getType, t1.getType) == Set(INT, LONG) =>
```

Comment:
avroSchema.getTypes doesn't return types, but rather some internal notion of types (e.g. org.apache.avro.Schema.IntType / LongType, etc.), but apparently they are protected or something, so I can't directly use them. However, if you call getType on them, you actually get Schema.Type.SOMETHING that you can use.

Oh, true, I didn't notice that. It appears though that you can't do match with sets in scala (unless you do some weird hacks). Or I am wrong :) Either way, I made this slightly shorter.